library/ssl_tls.c

```diff
@@ -1841,77 +1806,87 @@ static int ssl_decrypt_buf( mbedtls_ssl_context *ssl )
 
-            for( i = 0; i < ssl->transform_in->ivlen; i++ )
-                ssl->transform_in->iv_dec[i] = ssl->in_iv[i];
+            data += transform->ivlen;
+            rec->data_offset += transform->ivlen;
+            rec->data_len -= transform->ivlen;
+            rec->len[0] = (unsigned char)( rec->data_len >> 8 );
+            rec->len[1] = (unsigned char)( rec->data_len      );
         }
 #endif /* MBEDTLS_SSL_PROTO_TLS1_1 || MBEDTLS_SSL_PROTO_TLS1_2 */
 
-        if( ( ret = mbedtls_cipher_crypt( &ssl->transform_in->cipher_ctx_dec,
-                                   ssl->transform_in->iv_dec,
-                                   ssl->transform_in->ivlen,
-                                   dec_msg, dec_msglen,
-                                   dec_msg_result, &olen ) ) != 0 )
+        if( ( ret = mbedtls_cipher_crypt( &transform->cipher_ctx_dec,
+                                   transform->iv_dec, transform->ivlen,
+                                   data, rec->data_len, data, &olen ) ) != 0 )
         {
             MBEDTLS_SSL_DEBUG_RET( 1, "mbedtls_cipher_crypt", ret );
             return( ret );
         }
 
-        if( dec_msglen != olen )
+        if( rec->data_len != olen )
         {
             MBEDTLS_SSL_DEBUG_MSG( 1, ( "should never happen" ) );
             return( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
         }
 
 #if defined(MBEDTLS_SSL_PROTO_SSL3) || defined(MBEDTLS_SSL_PROTO_TLS1)
-        if( ssl->minor_ver < MBEDTLS_SSL_MINOR_VERSION_2 )
+        if( transform->minor_ver < MBEDTLS_SSL_MINOR_VERSION_2 )
         {
             /*
              * Save IV in SSL3 and TLS1
              */
-            memcpy( ssl->transform_in->iv_dec,
-                    ssl->transform_in->cipher_ctx_dec.iv,
-                    ssl->transform_in->ivlen );
+            memcpy( transform->iv_dec, transform->cipher_ctx_dec.iv,
+                    transform->ivlen );
         }
 #endif
-
-        padlen = 1 + ssl->in_msg[ssl->in_msglen - 1];
-
-        if( ssl->in_msglen < ssl->transform_in->maclen + padlen &&
-            auth_done == 0 )
+        if( rec->data_len == 0 )
         {
+            MBEDTLS_SSL_DEBUG_MSG( 1, ( "message doesn't contain padding byte" ) );
+            return( MBEDTLS_ERR_SSL_INVALID_MAC );
+        }
+
+        padlen = data[rec->data_len - 1];
 #if defined(MBEDTLS_SSL_DEBUG_ALL)
-            MBEDTLS_SSL_DEBUG_MSG( 1, ( "msglen (%d) < maclen (%d) + padlen (%d)",
-                        ssl->in_msglen, ssl->transform_in->maclen, padlen ) );
-#endif
-            padlen = 0;
-            correct = 0;
+        if( rec->data_len < padlen + 1 )
+        {
+            MBEDTLS_SSL_DEBUG_MSG( 1, ( "msglen (%d) < 1 + padlen (%d)",
+                                        rec->data_len, padlen ) );
         }
+#endif
+        correct *= ( rec->data_len >= padlen + 1 );
+        padlen  *= ( rec->data_len >= padlen + 1 );
+        /* Make sure padlen > 0 because we're later applying (_ % padlen) */
```

Comment:
You want to avoid `% secret_data` as the div instruction is rarely constant-time. See for example slide 26 of [this presentation](https://cryptojedi.org/peter/data/croatia-20160610.pdf#page=82) (the whole presentation is worth a read if you're not familiar with it yet).

Reply:
Note: I searched for `%.*padlen` in the file and could find no occurrence of `% padlen` except this comment.